src/gpio/wait_asynch.rs

```diff
@@ -0,0 +1,110 @@
+use core::{
+    future::Future,
+    sync::atomic::{AtomicBool, Ordering},
+    task::Poll,
+};
+extern crate alloc;
+use crate::gpio::{InputMode, InterruptType, Pin, PinDriver};
+use alloc::sync::Arc;
+use atomic_waker::AtomicWaker;
+use esp_idf_sys::{esp_nofail, gpio_intr_disable, EspError};
+
+impl<T: Pin, MODE: InputMode> embedded_hal_async::digital::Wait for PinDriver<'_, T, MODE> {
+    async fn wait_for_high(&mut self) -> Result<(), Self::Error> {
+        InputFuture::new(self, InterruptType::HighLevel)?.await;
+        Ok(())
+    }
+
+    async fn wait_for_low(&mut self) -> Result<(), Self::Error> {
+        InputFuture::new(self, InterruptType::LowLevel)?.await;
+        Ok(())
+    }
+
+    async fn wait_for_rising_edge(&mut self) -> Result<(), Self::Error> {
+        InputFuture::new(self, InterruptType::PosEdge)?.await;
+        Ok(())
+    }
+
+    async fn wait_for_falling_edge(&mut self) -> Result<(), Self::Error> {
+        InputFuture::new(self, InterruptType::NegEdge)?.await;
+        Ok(())
+    }
+
+    async fn wait_for_any_edge(&mut self) -> Result<(), Self::Error> {
+        InputFuture::new(self, InterruptType::AnyEdge)?.await;
+        Ok(())
+    }
+}
+
+pub(crate) struct InputFuture<'driver_ref, 'driver_struct, T: Pin, MODE: InputMode> {
+    // Unfortunately, the Wait trait uses functions that are given a mutable
+    // reference to the pin driver with no explicit lifetime parameter.  Since
+    // the reference has a different lifetime than the struct, we must make this
+    // explicit, or else the borrow checker will believe the reference must live
+    // forever.
+    driver: &'driver_ref mut PinDriver<'driver_struct, T, MODE>,
+    processed_interrupt: Arc<AtomicBool>,
+    atomic_waker: Arc<AtomicWaker>,
+}
+
+impl<'driver_ref, 'driver_struct, T: Pin, MODE: InputMode>
+    InputFuture<'driver_ref, 'driver_struct, T, MODE>
+{
+    pub(crate) fn new(
```

Comment:
Hmmmmmm. Futures should be inert (do nothing) until polled for the first time. In other words, shouldn't you call `subscribe` only when this future is polled for the first time? Are all other e-hal async drivers also doing this when the future is constructed?

Reply:
It's weird, right?

But they all do it. Or at least, the embassy ones do it (and they are pretty much the only ones that exist)
https://github.com/embassy-rs/embassy/blob/master/embassy-rp/src/gpio.rs

https://github.com/embassy-rs/embassy/blob/master/embassy-nrf/src/gpiote.rs#L380

https://github.com/embassy-rs/embassy/blob/master/embassy-stm32/src/exti.rs#L207

etc

Assuming someone thought about this, the main reason I would do it is because otherwise edge futures take a minimum of two polls to complete (first subscribes, second notices the edge).
The level ones can avoid this by checking for level on the first poll prior to setting up the interrupt.
That said, we don't have to be like this..

I agree with you that normally they should literally do nothing and you should be able to create and throw away futures without it having side-effects.

Reply:
> But they all do it. Or at least, the embassy ones do it (and they are pretty much the only ones that exist) 
> 
> Assuming someone thought about this, the main reason I would do it is because otherwise edge futures take a minimum of two polls to complete (first subscribes, second notices the edge). The level ones can avoid this by checking for level on the first poll prior to setting up the interrupt. That said, we don't have to be like this..

Oh well. Leave it like that for the first iteration.